pkg/component/controller/cplb/cplb_reconciler.go

```diff
@@ -109,22 +123,154 @@ func (r *CPLBReconciler) maybeUpdateIPs(endpoint *corev1.Endpoints) {
+			r.healthCheckers[addr].Stop()
+			delete(r.healthCheckers, addr)
+			shouldNotify = true
 		}
 	}
+	if shouldNotify {
+		r.updateCh <- struct{}{}
+	}
 }
 
 // GetIPs gets a thread-safe copy of the current list of IP addresses
 func (r *CPLBReconciler) GetIPs() []string {
 	r.mu.Lock()
 	defer r.mu.Unlock()
-	return slices.Clone(r.addresses)
+
+	var healthyAddrs []string
+	for addr := range r.healthCheckers {
+		if r.healthCheckers[addr].healthy {
+			healthyAddrs = append(healthyAddrs, addr)
+		}
+	}
+
+	if len(healthyAddrs) == 0 && len(r.healthCheckers) > 0 {
+		r.log.Warn("No healthy API servers detected by CPLBReconciler, returning every known address")
+		return slices.Collect(maps.Keys(r.healthCheckers))
+	}
+	return healthyAddrs
+}
+
+// healthChecker provides health checking functionality for Kubernetes API servers
+type healthChecker struct {
+	cancelFunc    context.CancelFunc
+	rc            rest.Interface
+	log           *logrus.Entry
+	healthy       bool
+	updateCh      chan<- struct{}
+	healthCheckFn func(context.Context) restResult
+}
+
+// newHealthChecker creates a new healthChecker instance
+func (r *CPLBReconciler) newHealthChecker(ctx context.Context, restConfig *rest.Config, addr string, healthCheckFn func(context.Context) restResult) *healthChecker {
+	var rc *rest.RESTClient
+	var err error
+
+	// In testing we don't provide a restConfig, therefore skip creating the rest client
+	if restConfig != nil {
+		cfg := rest.CopyConfig(restConfig)
+		cfg.GroupVersion = &corev1.SchemeGroupVersion
+		cfg.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+		cfg.Host = net.JoinHostPort(addr, strconv.Itoa(r.apiPort))
+
+		// Configure transport to establish new connections on every request
+		cfg.WrapTransport = func(rt http.RoundTripper) http.RoundTripper {
+			// If the original transport is an *http.Transport, modify it to disable connection reuse
+			if httpTransport, ok := rt.(*http.Transport); ok {
+				// Clone the transport to avoid modifying the original
+				transport := httpTransport.Clone()
+				transport.DisableKeepAlives = true // Disable HTTP keep-alive
+				transport.MaxIdleConns = 0         // No idle connections
+				transport.MaxIdleConnsPerHost = 0  // No idle connections per host
+				transport.IdleConnTimeout = 0      // Immediate timeout for idle connections
+				return transport
+			}
+			// If it's not an *http.Transport, return it as-is
+			return rt
+		}
+
+		rc, err = rest.RESTClientFor(cfg)
+		fmt.Println(rc.Client.Transport)
+	}
+
+	c, cancelFunc := context.WithCancel(ctx)
+	// If something went wrong or we didn't crate a rest client, run without health checking
+	if err != nil || rc == nil {
+		// I think this scenario is impossible outside testing.
+		r.log.WithError(err).Error(fmt.Sprintf("Unable to create restClient for %q healthCheck. Running without healthCheck", addr), err)
+		hc := &healthChecker{
+			healthy:    true,
+			cancelFunc: cancelFunc,
+		}
+		r.updateCh <- struct{}{}
```

Comment:
Could be non-blocking:

```suggestion
		select {
		case r.updateCh <- struct{}{}:
		default:
		}
```